### PR TITLE
refactor(typescript): refactor to ts (fixes #56)

### LINKED ISSUE #56: Port as-bind to TypeScript
related to #52 
related to #41 

We've had some issues with the typings. But all in all, it'd probably just be better to port everything to typescript for the type garuntees, and then just upload the TypeScript Source and things. :smile: 

In the meantime we will just commit the Typings sin to get things working in TypeScript projects :cry: (PRs welcome for a proper typings file! :smile: ) : https://www.stevefenton.co.uk/2013/01/complex-typescript-definitions-made-easy/

### PR DESCRIPTION
**Pull request by mathe42**
I wanted a look at the complete libary and took the chance to rewrite it to TS.

Can be optimized (for example TS-Strict mode is a good idea). But this results in a lot of errors.

### CODE REVIEW

**Review by torch2424 — COMMENTED** [review]
Woah! Thank you SOOOOOO much for doing this porting work! 

I'll also cc @surma again here, since it touches a lot of their code they implemented. But! I think you and I can handle this for the most part.

If you don't mind me asking, after doing the port, have you tried using the library from a non TS app? Like do all the same quick start examples and things work? 😄 I can test that if you would like! 

Let me know when you feel this PR is ready, I can do some testing on my end, and I think we can start considering merging this! Again, thank you SOOOOO much for doing this! 😄 🎉

File: lib/asbind-instance/type-converters.ts [hunk 1]
```diff
@@ -168,10 +207,10 @@ export function getConverterForType(typeName) {
   return { ascToJs: nop, jsToAsc: nop };
 }
 
-export function getAscToJsConverterForType(typeName) {
-  return getConverterForType(typeName)?.ascToJs;
+export function getAscToJsConverterForType(typeName: string) {
+  return getConverterForType(typeName).ascToJs;
```
[the remark this answers is not in the document]
Comment (mathe42) [reply lib/asbind-instance/type-converters.ts:211]: This was not needed before as well as `getConverterForType` always returns a non nullish value.

**Comment by mathe42** [comment]
> If you don't mind me asking, after doing the port, have you tried using the library from a non TS app? Like do all the same quick start examples and things work? 😄 I can test that if you would like!

Actualy I did not use it in a TS app 😉

Only type missmatch I could think of is in https://github.com/torch2424/as-bind/blob/3ff4b1d82d369abd874969271649765f1606dd16/lib/types.ts#L3-L7 I'm not shure with that...

**Comment by mathe42** [comment]
> Let me know when you feel this PR is ready, I can do some testing on my end, and I think we can start considering merging this! Again, thank you SOOOOO much for doing this! 😄 🎉

Ready for review.

**Comment by torch2424** [comment]
> This was not needed before as well as getConverterForType always returns a non nullish value.

Sounds good to me! 😄 

> Actualy I did not use it in a TS app wink only a simple JS

Yay! Awesome, thank you! 😄 🎉 

> Ready for review.

Awesome! So since there wasn't any changes since the last time I looked at the code, I am going to approve and merge! 😄

**Review by torch2424 — APPROVED** [review]
I also went ahead and tested this in VSCode, and it worked great! 😄 🎉 

Screenshot from 2021-09-29 10-19-43

**Comment by torch2424** [comment]
@mathe42 I am going to merge in all your changes, and then cut a minor release after I review and merge all of your code 😄

**Pull request merged**